src/api.cc

```diff
@@ -119,57 +126,87 @@ AkliteClient::~AkliteClient() {
+static std::vector<TufTarget> filterTargets(const std::vector<Uptane::Target> allTargets,
+                                            const Uptane::HardwareIdentifier hwidToFind,
+                                            const std::vector<std::string>& tags,
+                                            const std::vector<std::string>& secondary_hwids) {
   std::vector<TufTarget> targets;
-  for (const auto& t : client_->allTargets()) {
+  for (const auto& t : allTargets) {
     int ver = 0;
     try {
       ver = std::stoi(t.custom_version(), nullptr, 0);
     } catch (const std::invalid_argument& exc) {
       LOG_ERROR << "Invalid version number format: " << t.custom_version();
       ver = -1;
     }
-    if (!target_has_tags(t, client_->tags)) {
+    if (!target_has_tags(t, tags)) {
       continue;
     }
     for (const auto& it : t.hardwareIds()) {
       if (it == hwidToFind) {
         targets.emplace_back(t.filename(), t.sha256Hash(), ver, t.custom_data());
         break;
       }
-      for (const auto& hwid : secondary_hwids_) {
+      for (const auto& hwid : secondary_hwids) {
         if (it == Uptane::HardwareIdentifier(hwid)) {
           targets.emplace_back(t.filename(), t.sha256Hash(), ver, t.custom_data());
           break;
         }
       }
     }
   }
-
   std::sort(targets.begin(), targets.end(), compareTargets);
-  return CheckInResult(status, client_->config.provision.primary_ecu_hardware_id, targets);
+  return targets;
+}
+
+CheckInResult AkliteClient::UpdateMetaAndGetTargets(std::shared_ptr<aklite::tuf::RepoSource> repo_src) const {
+  auto status = CheckInResult::Status::Ok;
+
+  LOG_INFO << "Refreshing Targets metadata";
+  try {
+    tuf_repo_->updateMeta(repo_src);
+  } catch (const std::exception& e) {
+    LOG_WARNING << "Unable to update latest metadata, using local copy: " << e.what();
+    try {
+      tuf_repo_->checkMeta();
+    } catch (const std::exception& e) {
+      LOG_ERROR << "Unable to use local copy of TUF data";
+      return CheckInResult(CheckInResult::Status::Failed, "", {});
+    }
+    status = CheckInResult::Status::OkCached;
+  }
+
+  auto allTargetsTuf = tuf_repo_->GetTargets();
+  std::vector<Uptane::Target> allTargets;
+  for (auto const& ut : allTargetsTuf) {
+    allTargets.emplace_back(Target::fromTufTarget(ut));
+  }
+
+  Uptane::HardwareIdentifier hwidToFind(client_->config.provision.primary_ecu_hardware_id);
+  auto matchingTargets = filterTargets(allTargets, hwidToFind, client_->tags, secondary_hwids_);
+  return CheckInResult(status, client_->config.provision.primary_ecu_hardware_id, matchingTargets);
+}
+
+CheckInResult AkliteClient::CheckIn() const {
+  if (!configUploaded_) {
+    client_->reportAktualizrConfiguration();
+    configUploaded_ = true;
+  }
+  client_->reportNetworkInfo();
+  client_->reportHwInfo();
+  client_->reportAppsState();
+
+  boost::property_tree::ptree pt;
+  pt.put<std::string>("tag", client_->tags.empty() ? "" : client_->tags.at(0));
```

Comment:
I think we need to set three more headers here: `x-ats-dockerapps`, `x-ats-target`, `x-ats-ostreehash` to make Device Gateway happy. The `checkin` function (https://github.com/foundriesio/device-gateway/blob/f1008143f27e6dc11dbf5bcf718d759a3df343d4/server/handlers_meta.go#L105) is called by the root meta handler, and these headers' values are used to set a device state at the DB.

Obviously, in the future when the tufctl becomes a separate/standalone daemon we will need to figure out how to change/adjust overall approach since the tufctl service won't know or shouldn't know about the currently installed target, ostree and list of configured apps.

Reply:
I've now added the fields to `CheckIn` logic, which was missing in the previous patches.

Reply:
ok. I tested it by using the custom sota client/example, and it works fine now, a device status is not "erased" as it is used to be.